opm/input/eclipse/Schedule/MSW/SegmentMatcher.hpp

```diff
@@ -0,0 +1,303 @@
+        using WellSegmentRangeIterator = std::vector<int>::const_iterator;
+
+        /// Segment Range for Single MS Well.
+        class WellSegmentRange
+        {
+        public:
+            /// Start of Range
+            WellSegmentRangeIterator begin() const { return this->begin_; }
+
+            /// End of Range
+            WellSegmentRangeIterator end() const { return this->end_; }
+
+            friend class SegmentSet;
+
+        private:
+            /// Start of Range
+            WellSegmentRangeIterator begin_{};
+
+            /// End of Range
+            WellSegmentRangeIterator end_{};
+
+            /// Default Constructor.
+            ///
+            /// Empty range.
+            ///
+            /// For use by SegmentSet only.
+            WellSegmentRange() = default;
+
+            /// Non-Empty Range
+            ///
+            /// For use by SegmentSet only.
+            ///
+            /// \param[in] begin Start of range.
+            /// \param[in] end End of range.
+            WellSegmentRange(WellSegmentRangeIterator begin,
+                             WellSegmentRangeIterator end)
+                : begin_{ begin }
+                , end_  { end }
+            {}
+        };
+
+        /// Default Constructor.
+        SegmentSet();
+
+        /// Predicate for whether or not segment set is empty.
+        ///
+        /// \return Whether or not segment set is empty.
+        bool empty() const
+        {
+            return this->segments_.empty();
+        }
+
+        /// Predicate for whether or not segment set applies to a single
+        /// segment in a single MS well.
+        ///
+        /// \return Whether or not segment set is a single segment in a
+        ///   single MS well.  Useful to distinguish whether or not this
+        ///   segment set generates a scalar UDQ or a UDQ set in the context
+        ///   of a segment level UDQ.
+        bool isScalar() const
+        {
+            return this->segments_.size() == std::vector<int>::size_type{1};
+        }
+
+        /// Retrieve list of (MS) well names covered by this result set.
+        ///
+        /// \return List MS well names covered by this result set.
+        std::vector<std::string_view> wells() const;
+
+        /// Retrive result set's segments for single MS well.
+        ///
+        /// \param[in] well Named MS well.  Should usually be one of the
+        ///    items in the return value from \code wells() \endcode.
+        ///
+        /// \return range of \c well's segments matching the input request.
+        ///    Empty unless \p well is one of the return values from \code
+        ///    wells() \endcode.
+        WellSegmentRange segments(std::string_view well) const;
+
+        friend class SegmentMatcher;
+
+    private:
+        /// List of MS wells covered by this result set.
+        std::vector<std::string> wells_{};
+
+        /// Sorted index into \c wells_ for faster lookup.
```

Comment:
How this is sorted is not really clear unless you read the code or the tests. Please clarify

Reply:
> How this is sorted is not really clear unless you read the code or the tests

That's actually intentional.  The user should not need to know (or care) how we form lookup indices if they're "just" looking at the result of a matching process.

Reply:
Sorry, but for people working with the code (and reviewers) this is very important information. Not documenting it, just costs everybody besides you more time when working the code. Hence I consider this ( x time N minutes instead of one time n minutes with N>>n) a bad intention.